core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -84,9 +74,130 @@ abstract class InterBrokerSendThread(name: String,
     }
   }
 
-  def wakeup(): Unit = networkClient.wakeup()
+  private def sendRequests(now: Long): Unit = {
+    var pollTimeout = Long.MaxValue
+    if (unsentRequests.hasRequests) {
+      for (node <- unsentRequests.nodes.asScala) {
+        val requestIterator = unsentRequests.requestIterator(node)
+        while (requestIterator.hasNext) {
+          val request = requestIterator.next
+          if (networkClient.ready(node, now)) {
+            networkClient.send(request, now)
+            requestIterator.remove()
+          } else
+            pollTimeout = if (unsentRequests.hasRequests) AdminClient.DefaultReconnectBackoffMs else Long.MaxValue
+        }
+      }
+    }
+    networkClient.poll(pollTimeout, now)
+  }
+
+  private def checkDisconnects(now: Long): Unit = {
+    // any disconnects affecting requests that have already been transmitted will be handled
+    // by NetworkClient, so we just need to check whether connections for any of the unsent
+    // requests have been disconnected; if they have, then we complete the corresponding future
+    // and set the disconnect flag in the ClientResponse
+    val nodeIterator = unsentRequests.nodeIterator()
```

Comment:
nit: wouldn't it be better to iterate the map entries instead of the keys?

Reply:
+1